src/main/scala/util/RunUtils.scala

```diff
@@ -393,134 +623,76 @@ object RunUtils {
-                  val newCall = GoTo(newBlocks, indirectCall.label)
-                  block.replaceJump(newCall)
-                }
-              case LiftedBottom =>
-              }
-          case _ =>
-      case _ =>
-    }
+    IRTransform.doCleanup(ctx)
 
-    def nameExists(name: String): Boolean = {
-      IRProgram.procedures.exists(_.name.equals(name))
-    }
+    q.loading.dumpIL.foreach(s => writeToFile(serialiseIL(ctx.program), s"$s-before-analysis.il"))
+    val analysis = q.staticAnalysis.map(conf => staticAnalysis(conf, ctx))
+    q.loading.dumpIL.foreach(s => writeToFile(serialiseIL(ctx.program), s"$s-after-analysis.il"))
 
-    def addFakeProcedure(name: String): Unit = {
-      IRProgram.procedures += Procedure(name)
+    if (q.runInterpret) {
+      val interpreter = Interpreter()
+      interpreter.interpret(ctx.program)
     }
 
-    def resolveAddresses(valueSet: Set[Value]): Set[AddressValue] = {
-      var functionNames: Set[AddressValue] = Set()
-      valueSet.foreach {
-        case globalAddress: GlobalAddress =>
-          if (nameExists(globalAddress.name)) {
-            functionNames += globalAddress
-            Logger.info(s"RESOLVED: Call to Global address ${globalAddress.name} resolved.")
-          } else {
-            addFakeProcedure(globalAddress.name)
-            functionNames += globalAddress
-            Logger.info(s"Global address ${globalAddress.name} does not exist in the program.  Added a fake function.")
-          }
-        case localAddress: LocalAddress =>
-          if (nameExists(localAddress.name)) {
-            functionNames += localAddress
-            Logger.info(s"RESOLVED: Call to Local address ${localAddress.name}")
-          } else {
-            addFakeProcedure(localAddress.name)
-            functionNames += localAddress
-            Logger.info(s"Local address ${localAddress.name} does not exist in the program. Added a fake function.")
-          }
-        case _ =>
+    IRTransform.prepareForTranslation(q.loading, ctx)
+
+    Logger.info("[!] Translating to Boogie")
+    val boogieTranslator = IRToBoogie(ctx.program, ctx.specification)
+    val boogieProgram = boogieTranslator.translate(q.boogieTranslation)
+
+    BasilResult(ctx, analysis, boogieProgram)
+  }
+
+  /** Use static analysis to resolve indirect calls and replace them in the IR until fixed point.
+    */
+  def staticAnalysis(config: StaticAnalysisConfig, ctx: IRContext): StaticAnalysisContext = {
+    var iteration = 1
+    var modified: Boolean = true
+    var analysisResult = mutable.ArrayBuffer[StaticAnalysisContext]()
+    while (modified) {
+      var newIR = null
+      Logger.info("[!] Running Static Analysis")
+      val result = StaticAnalysis.analyse(ctx, config, iteration)
+      analysisResult.append(result)
+      Logger.info("[!] Replacing Indirect Calls")
+      val (ir, mod) = IRTransform.resolveIndirectCalls(result.cfg, result.vsaResult, ctx.program)
```

Comment:
ir is never used for anything? Which means further iterations won't actually build upon the resolved indirect calls?

Reply:
@ailrst

Reply:
The IR is modified in-place, the top-level 'program' is not copied or replaced in any way.